### PR TITLE
Tuples!

### PR DESCRIPTION
**Pull request by hypercubestart**
Included in this PR:
- new glenside language constructs
  - ConstructTuple
  - TupleGetItem
- additional Relay Opaque Operators
  - RelayLeakyReLU
  - RelayAvgPool2D
  - RelayUpSampling
  - RelaySigmoid
  - RelayMaximum
  - RelayMinimum
- edit `fn create_worklist` in `from_relay/mod.rs` to efficiently parse larger programs
- edit `run_relay.py` to allow tuple output

### CODE REVIEW

File: src/language/from_relay/mod.rs [hunk 1]
```diff
@@ -662,22 +674,59 @@ fn compile_expression(
...
-                .upcast::<tvm::ir::expr::BaseExpr>()
-                .downcast::<tvm::ir::op::Op>()
-                .unwrap()
-                .name
-                .as_str(),
-            Ok("nn.batch_norm")
-        );
-        assert_eq!(tuple_get_item.index, 0);
+            .downcast::<tvm::ir::relay::Call>();
+
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:678]: There's a comment a few lines up ([line 670](https://github.com/gussmith23/glenside/pull/94/files#diff-662d20b9cf2210216fea2642a0736aa1edf391b04248d7b90210a09199df78ebR670)) that should be updated, because it's no longer true!

Recapping my own understanding: the tuple_get_item code looks this way because I used to only parse tuple_get_item in the case where it was a tuple_get_item getting the first item of the result of a `nn.batch_norm`, in which case I just compiled it straight to a `BatchNormInference`. This is a hack, plain and simple, but it's a useful hack, and is still a useful hack, so we need to keep it around. At the same time, we need to make the code more general, now that we actually want to compile tuples.

Here's what I think we need: we should keep code for the special case where we have a tuple_get_item accessing index 0 of a batch norm, and we should have that compile to a Glenside RelayOperator batch norm (as currently happens), though we should hide it behind the `simplify_batch_norm_for_inference` argument that gets passed in to from_relay. This feature flag is also checked in the code that compiles an `nn.batch_norm` expression. If the flag is set, that code will know to compile an `nn.batch_norm` to a single tensor output; if it's not set, we should panic (because we haven't yet implemented the full Relay batch norm). That's already done, though. It seems like you do all of this already, _except_ you need to add a check for `simplify_batch_norm_for_inference_hack`, maybe in the if condition on line 681?

File: src/language/from_relay/mod.rs [hunk 2]
```diff
@@ -1472,6 +1566,367 @@ fn compile_expression(
...
+                    for i in 0..attrs.newshape.len() {
+                        let val = attrs
+                            .newshape
+                            .get(i.try_into().ok().unwrap())
+                            .unwrap()
+                            .downcast::<IntImm>()
+                            .unwrap()
+                            .value;
+                        new_shape.push(if val < 0 {
+                            // assume -1 = 1 in reshape.newshape
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:1658]: Is this a safe assumption? Are there checks we can add here so that we panic when the assumption is not true?

File: src/language/language.rs [hunk 3]
```diff
@@ -305,6 +313,9 @@ pub enum RelayOperator {
     /// (relay-operator relay-relu <data: access>)
     RelayReLU,
 
+    /// (relay-operator relay-relu <data: access> <data: access>)
```
Comment (gussmith23) [inline src/language/language.rs:316]:
```suggestion
    /// (relay-operator relay-relu <data: access> <some-other-name: access>)
```
The parameters have identical names

File: src/language/language.rs [hunk 4]
```diff
@@ -523,6 +563,7 @@ pub enum MyAnalysisData {
     Legacy(MyAnalysisDataLegacyData),
     AccessPattern(AccessPatternData),
     Shape(ShapeData),
+    Tuple(Vec<Box<MyAnalysisData>>),
```
Comment (gussmith23) [inline src/language/language.rs:566]: Actually, I know I suggested `Vec<Box<...>>`, but a `Vec<MyAnalysisData>` might work just fine -- can you try that?

Comment (gussmith23) [reply src/language/language.rs:566]: BTW, the whole point of this is that you can't have directly recursive data structures in Rust, as Rust will have no idea what the stack size of the data structure is at compile time. But a Vec is allocated on the heap, so I'm pretty sure its stack size is known at compile time. I could be wrong.

File: src/language/language.rs [hunk 5]
```diff
@@ -5206,4 +5417,53 @@ mod tests {
...
+        let mut egraph =
+            egg::EGraph::<Language, MyAnalysis>::new(MyAnalysis { name_to_shape: map });
+        let id = egraph.add_expr(&program);
+        match &egraph[id].data {
+            MyAnalysisData::AccessPattern(b) => {
+                assert_eq!(b.shape, IxDyn(&[16, 32]));
+            }
+            _ => panic!(),
+        }
+    }
```
Comment (gussmith23) [inline src/language/language.rs:5468]: Still need tests for other new language constructs (I realize this PR is WIP!)

**Review by gussmith23 — CHANGES_REQUESTED** [review]
First pass!

File: Cargo.toml [hunk 6]
```diff
@@ -39,8 +39,8 @@ git = "https://github.com/gussmith23/rplex"
 # issue.
 # NOTE Keep glenside-evaluation in sync with this
 # If the versions get out of whack, we'll probably have some weird errors.
-rev = "ff5450f9cbe07d57e73594c5fbd68f3bd665c2dc"
-git = "https://github.com/gussmith23/tvm"
+rev = "4ed759f6b6cafbab707f95b57762913a1f57c021"
+git = "https://github.com/hypercubestart/incubator-tvm"
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply Cargo.toml:43]: not completely, last thing that needs to be merged is this change from Jared: https://github.com/apache/tvm/pull/7085

File: src/codegen.rs [hunk 7]
```diff
@@ -1753,31 +1772,89 @@ if (i{i} < {dim_len}) {{
...
-    use ndarray_npy::{read_npy, write_npy};
-    use ndarray_rand::{rand_distr::Uniform, RandomExt};
-    use rand::{rngs::SmallRng, Rng, SeedableRng};
-    use std::fs::File;
-    use std::io::Write;
-    use std::iter::FromIterator;
-    use std::path::PathBuf;
-    use std::process::Command;
-    use std::str::FromStr;
+pub fn run_relay(
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:1775]: good catch! this was from back when I was debugging and needed to create an external test

File: src/codegen.rs [hunk 8]
```diff
@@ -4042,4 +4141,295 @@ int main() {{
...
+
+        let mut egraph = EGraph::new(MyAnalysis {
+            name_to_shape: env.clone(),
+        });
+
+        let id = egraph.add_expr(&expr);
+
+        let result = run_relay_tuple_out(&value_env, &shapes_vec, &relay, 12);
+
+        // let code = codegen(
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:4193]: yep you got it, codegen is still missing

File: src/codegen.rs [hunk 9]
```diff
@@ -4042,4 +4141,295 @@ int main() {{
...
+
+        let mut egraph = EGraph::new(MyAnalysis {
+            name_to_shape: env.clone(),
+        });
+
+        let id = egraph.add_expr(&expr);
+
+        let result = run_relay(&value_env, &shapes_vec, &relay);
+
+        // let code = codegen(
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/codegen.rs:4339]: good idea! ill do this

File: src/language/from_relay/mod.rs [hunk 10]
```diff
@@ -464,10 +464,10 @@ pub fn shape_from_type(t: tvm::ir::ty::Type) -> Vec<usize> {
...
             )
         });
-    assert_eq!(
-        tensor_type.dtype.clone(),
-        "float32".parse().unwrap(),
-        "only supporting float32x1 at the moment"
+    assert!(
+        tensor_type.dtype.clone() == "float32".parse().unwrap()
+            || tensor_type.dtype.clone() == "int32".parse().unwrap(),
+        "only supporting float32x1 and in32x1 at the moment"
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/language/from_relay/mod.rs:470]: done! thanks!

File: src/language/from_relay/mod.rs [hunk 11]
```diff
@@ -547,28 +548,37 @@ pub fn from_relay(
 /// so we first recursively generate a worklist which we can then iterate over.
 /// The main goal of the worklist is to make sure an expression comes *after*
 /// its children in the worklist; otherwise, we can't compile the expression!
-fn create_worklist(relay_expr: Expr, worklist: &mut Vec<Expr>) {
-    fn add_to_worklist(expr: Expr, worklist: &mut Vec<Expr>) {
-        if !worklist.contains(&expr) {
-            worklist.push(expr.clone());
-        }
+fn create_worklist(relay_expr: Expr, worklist: &mut Vec<Expr>, visited: &mut HashSet<Expr>) {
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/language/from_relay/mod.rs:551]: done!

File: src/language/from_relay/mod.rs (hunk 2 above)
Comment (hypercubestart) [reply src/language/from_relay/mod.rs:1658]: For the models we use this assumption is true.

I think the tricky engineering bit would be to figure out how to infer the shapes since we need the shape information of its first argument. AFAIK shape information is only available after the expr is added to the egraph.

File: src/language/language.rs (hunk 3 above)
Comment (hypercubestart) [reply src/language/language.rs:316]: oops this is supposed to be alpha!

File: src/language/language.rs (hunk 4 above)
Comment (hypercubestart) [reply src/language/language.rs:566]: ah, I thought the problem was with enums of different sizes but it turns out all instantiations of an enum have the same size

File: Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (gussmith23) [reply Cargo.toml:43]: Great, can you open an issue to point TVM back to the main repo, and give some context? Say that it was changed to point to your fork in this PR, and that it's waiting on that PR

File: src/language/from_relay/mod.rs (hunk 2 above)
Comment (gussmith23) [reply src/language/from_relay/mod.rs]: Like we discussed, the shape information *should* be available via `checked_type`. Let me know if that's not the case!

Comment (hypercubestart) [reply src/language/from_relay/mod.rs]: ah sorry the new commits aren't reflected. I decided to just use the shape information via `checked_type` instead of computing it ourselves. lmk if this is okay!

File: src/language/from_relay/mod.rs [hunk 12]
```diff
@@ -655,29 +671,61 @@ fn compile_expression(
...
-                .clone()
-                .upcast::<tvm::ir::expr::BaseExpr>()
-                .downcast::<tvm::ir::op::Op>()
-                .unwrap()
-                .name
-                .as_str(),
-            Ok("nn.batch_norm")
-        );
-        assert_eq!(tuple_get_item.index, 0);
+            .downcast::<tvm::ir::relay::Call>();
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:683]: The `tuple` field might not be a `Call` anymore! For example, if we have something like `(<something>, <something else>).1`, then the `tuple` field of the `tuple_get_item` will be a tuple constructor.

This can be fixed *and* the logic of this `else if` branch can be simplified a lot, all at once. I think the following structure makes sense:
First, check if `simplify_batch_norm_for_inference_hack` is true && if the `tuple` is a batch norm && if `tuple_get_item.index == 0`. If all three of those things are true, then we should apply the hack. The hack, in this case, is the fact that we don't create a `TupleGetItem`, but instead, we directly return the batch norm (`get_compiled_expression(tuple_get_item.tuple.clone())`).
If the hack doesn't fire, then just generate the `TupleGetItem` as you already do on lines 707-710.

Can you check my logic on that? I think it makes sense, but it might not.

File: src/language/language.rs [hunk 13]
```diff
@@ -1641,6 +1742,117 @@ impl egg::Analysis<Language> for MyAnalysis {
...
+                                    layout.clone(),
+                                    crate::language::RelayActivationLayout::NCHW,
+                                    "upsampling only supports NCHW"
+                                );
+                                // let mut shape = array![a.shape[0], a.shape[1], scale_h.into() * shape[2], scale_w.into() * shape[w]];
+                                let mut shape = a.shape.clone();
+                                assert_eq!(scale_h.ndim(), 0);
+                                assert_eq!(scale_w.ndim(), 0);
+                                shape[2] = (scale_h.first().unwrap() * (shape[2] as f64)) as usize;
+                                shape[3] = (scale_w.first().unwrap() * (shape[3] as f64)) as usize;
```
Comment (gussmith23) [inline src/language/language.rs:1837]: Does the right kind of rounding happen here? What happens if the floats are not integers?

**Review by gussmith23 — COMMENTED** [review]
More comments. Also waiting on tests in `from_relay`, when you get the chance!

File: src/language/language.rs (hunk 13 above)
Comment (hypercubestart) [reply src/language/language.rs:1837]: it should fail!

File: src/language/language.rs [hunk 14]
```diff
@@ -313,7 +313,7 @@ pub enum RelayOperator {
     /// (relay-operator relay-relu <data: access>)
     RelayReLU,
 
-    /// (relay-operator relay-relu <data: access> <data: access>)
+    /// (relay-operator relay-relu <data: access> <alpha: Float64>)
```
[the remark this answers is not in the document]
Comment (hypercubestart) [reply src/language/language.rs:316]: good catch!

File: src/language/from_relay/mod.rs (hunk 12 above)
Comment (hypercubestart) [reply src/language/from_relay/mod.rs:683]: yep that makes sense! im realizing now that we may run into a little trouble if `tuple_get_item.tuple` isn't a `Op`.

Also, I have duplicated code here to comply with Rust's type checking. If there's any simpler way to do this lmk!

File: Cargo.toml (hunk 6 above)
[the remark this answers is not in the document]
Comment (hypercubestart) [reply Cargo.toml:43]: haha unsure if this is what you wanted, but I found an old open issue that we could reuse: https://github.com/gussmith23/glenside/issues/47

File: src/codegen.rs [hunk 15]
```diff
@@ -4042,4 +4144,107 @@ int main() {{
                 .expect("Could not convert stderr to UTF8")
         );
     }
+
+    #[test]
+    #[should_panic(expected = "unfinished test")]
```
Comment (gussmith23) [inline src/codegen.rs:4149]: Sorry for flip-flopping on this, but could you actually change this (and the other unimplemented tests) to
```suggestion
    #[ignore = "unfinished test"]
```
It will help me keep track of tests that need to be changed in the long-term. The #should_panic strategy is good for the short term, while you're working on a PR, but I don't want to check in code that hides unfinished tests without calling them out in some way. Using #ignore will at least make them show up as ignored when I run the test suite.

File: src/codegen.rs [hunk 16]
```diff
@@ -1043,6 +1051,12 @@ add_with_broadcasting((float*) {out}, (float*) {X}, (float*) {Y}, (int*)  {out_s
 
                     Some(add_out)
                 }
+                RelayOperator::RelayLeakyReLU => todo!(),
```
Comment (gussmith23) [inline src/codegen.rs:1054]: Can you add `#[ignore]`d tests in this module for these new operators? (See comment below about `#[ignore]`.) You can just copy and paste the tests for the other operators and modify them!

File: src/language/from_relay/mod.rs [hunk 17]
```diff
@@ -655,29 +671,51 @@ fn compile_expression(
...
+                get_compiled_expression(tuple_get_item.tuple.clone())
+            } else {
+                let data_id = get_compiled_expression(tuple_get_item.tuple.clone());
+                let index_id = glenside_expr.add(Language::Usize(tuple_get_item.index as usize));
+                glenside_expr.add(Language::TupleGetItem([data_id, index_id]))
+            }
+        } else {
+            let data_id = get_compiled_expression(tuple_get_item.tuple.clone());
+            let index_id = glenside_expr.add(Language::Usize(tuple_get_item.index as usize));
+            glenside_expr.add(Language::TupleGetItem([data_id, index_id]))
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:707]: This logic is duplicated. If you delete 700-702, I'm fairly certain this code will do exactly the same thing, because we'll just fall through to the outer `else` branch (the branch on 705-707). Can you check my thinking on that? I'd rather not duplicate logic if possible!

File: src/language/from_relay/mod.rs [hunk 18]
```diff
@@ -655,29 +671,51 @@ fn compile_expression(
...
+                    == "nn.batch_norm"
+                && tuple_get_item.index == 0
+            {
+                get_compiled_expression(tuple_get_item.tuple.clone())
+            } else {
+                let data_id = get_compiled_expression(tuple_get_item.tuple.clone());
+                let index_id = glenside_expr.add(Language::Usize(tuple_get_item.index as usize));
+                glenside_expr.add(Language::TupleGetItem([data_id, index_id]))
+            }
+        } else {
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:704]: Can you add a simple comment here, saying that this is the standard case, where Relay `TupleGetItem`s are converted directly to Glenside `TupleGetItem`s? I just want to make it clear that the `if` branch is a corner case.

File: src/language/from_relay/mod.rs [hunk 19]
```diff
@@ -809,6 +847,29 @@ fn compile_expression(
                         compute(glenside_expr, ComputeType::ReLU, data_id)
                     }
                 }
```
Comment (gussmith23) [inline src/language/from_relay/mod.rs:849]: Can you add tests for all of the new Relay constructs we support here? I recognize that some of them we can't currently test because codegen is not implemented for them -- these you can just mark with an `#[ignore = "..."]` (see comment above). But some of these can be tested, like concatenate, reshape, split, transpose, and squeeze (there might be others).

File: src/language/from_relay/run_relay.py [hunk 20]
```diff
@@ -12,13 +12,12 @@
...
                     type=argparse.FileType('rb'),
                     default=sys.stdin)
-parser.add_argument('npy_out_filepath', type=argparse.FileType('wb'))
-parser.add_argument('npy_arg_filepath',
-                    nargs='*',
-                    type=argparse.FileType('rb'))
+parser.add_argument('--npy_out_filepath', 
+                    nargs='*')
+parser.add_argument('--npy_arg_filepath',
+                    nargs='*')
```
Comment (gussmith23) [inline src/language/from_relay/run_relay.py:20]: Is it not possible to use `type=...` with flag-type arguments in `argparse`? Why did you remove the `type=argparse.FileType(...)`?

File: src/language/from_relay/run_relay.py [hunk 21]
```diff
@@ -31,7 +30,20 @@
...
+inputs = []
+for filepath in parsed.npy_arg_filepath:
+    with open(filepath, 'rb') as arg_file:
+        inputs.append(np.load(arg_file))
 
-np.save(parsed.npy_out_filepath, output)
+# need graph runtime or crashes for yolo
+output = relay.create_executor(mod=expr, kind="graph").evaluate()(*inputs)
+
+if type(output) is list:
```
Comment (gussmith23) [inline src/language/from_relay/run_relay.py:41]: Can you add an assert here, checking that the length of the Relay output is the same as the number of output filepaths given to the program?

File: src/language/language.rs [hunk 22]
```diff
@@ -314,6 +325,15 @@ pub enum RelayOperator {
     ///  <layout: RelayActivationLayout>)
     RelayGlobalAvgPool2D,
 
+    /// (relay-operator relay-avg-pool2d <data: access>
+    /// <pool_size: shape> <strides: shape> <padding: shape> <layout: RelayActivationLayout>
+    /// )
+    RelayAvgPool2D,
+
+    /// (relay-operator relay-upsampling <data: access> <scale_h: Float64> <scale_w: Float64>
+    /// <layout: RelayActivationLayout>
```
Comment (gussmith23) [inline src/language/language.rs:334]:
```suggestion
    /// <layout: RelayActivationLayout>)
```

File: src/language/language.rs (hunk 13 above)
Comment (gussmith23) [reply src/language/language.rs:1837]: I'm unsure about this. The documentation for it in Relay implies there's more complexity here: https://tvm.apache.org/docs/api/python/relay/nn.html#tvm.relay.nn.upsampling. I think they would have made the parameters integers, if they could only be integers. For now, can we just assert that the float values must be integers? And then in the future we can support fractional values if we need to.

Comment (gussmith23) [reply src/language/language.rs:1837]: I looked in to the TVM implementation a bit more. See [this code in TOPI](https://github.com/apache/tvm/blob/da446af3b73a9b7354b89bb5792968cd705359d8/python/tvm/topi/nn/upsampling.py#L67). The scales are multiplied as-is (i.e., with the fractional part) with the size of the image to calculate the new size of the image. So I think they can be fractional!

**Review by gussmith23 — COMMENTED** [review]
Some more comments! Nearly there. These comments are mostly about tests.

File: src/language/from_relay/mod.rs (hunk 17 above)
Comment (hypercubestart) [reply src/language/from_relay/mod.rs:707]: hmm I gave it a try and RUst's type checker didn't like it. It looks its of the if expression, so the inner if-else expression needs type Egg::Id, and if the lines on 700-702 are deleted, Rust implicitly has a unit type

File: src/language/from_relay/run_relay.py (hunk 20 above)
Comment (hypercubestart) [reply src/language/from_relay/run_relay.py:20]: ah, it was because when I was testing yolo, the test failed on the argument parser.

It turns out that using the `type=...` argument opens all the files at once but having to open so many files at the same time without closing any caused the test to fail because theres a limit to the number of open files at once :p

I'll add a comment to this

File: src/language/language.rs (hunk 13 above)
Comment (hypercubestart) [reply src/language/language.rs:1837]: hmm it looks like right after that part of the code they calculate the new resized_shape https://github.com/apache/tvm/blob/da446af3b73a9b7354b89bb5792968cd705359d8/python/tvm/topi/nn/upsampling.py#L70

                simplify(topi.cast(te.round(scaled_h), data.shape[2].dtype)),
                simplify(topi.cast(te.round(scaled_w), data.shape[3].dtype)),

which looks like it rounds the scaled numbers to the nearest integer if I am reading this right?

File: src/language/from_relay/mod.rs (hunk 19 above)
Comment (hypercubestart) [reply src/language/from_relay/mod.rs:854]: added tests for all the new Relay constructs added in this PR! i commented out those that fail because of the issue with using #[ignore]

File: src/language/language.rs (hunk 13 above)
Comment (gussmith23) [reply src/language/language.rs:1837]: I guess the question is: does `te.round()` round the same as `as usize`? If that's the case, then yeah, your code is right.

Can you use [round](https://doc.rust-lang.org/std/primitive.f64.html#method.round) instead of just `as usize`? That will bring us a little closer to the behavior, I think. Also, can you add a TODO saying that the rounding may not be exactly the same as TVM's?

Honestly, there's no good answer if we want to be exact here, and I'm just going to bank on it not being a problem!

Comment (hypercubestart) [reply src/language/language.rs]: added round!

File: src/language/from_relay/mod.rs (hunk 17 above)
Comment (gussmith23) [reply src/language/from_relay/mod.rs]: I see. I changed it -- I just used a `return` to short-circuit.

File: src/language/from_relay/run_relay.py (hunk 20 above)
Comment (gussmith23) [reply src/language/from_relay/run_relay.py:20]: Haha, what a bug. Good catch. Thanks!

File: src/language/from_relay/mod.rs (hunk 12 above)
Comment (gussmith23) [reply src/language/from_relay/mod.rs]: Simplified this code a bit, we're good here now!

**Review by gussmith23 — APPROVED** [review]
I'll merge this once I make sure the tests pass! Thanks so much for all of your hard work!!

**Comment by gussmith23** [comment]
@hypercubestart hmm...any idea where this new error would be coming from? https://github.com/gussmith23/glenside/pull/94/checks?check_run_id=1866343438

It seems pretty dumb...it seems like somehow [this line](https://github.com/apache/tvm/blob/2999d03284c74f6840503ae3b880d3579a76f1af/python/tvm/autotvm/tophub.py#L184) is running, even though the path already exists. It's really odd to me that only a single test is failing with this error, too.

**Comment by hypercubestart** [comment]
ah I think i ran into that before, and rerunning the tests worked

**Comment by gussmith23** [comment]
🤞 hope so! i already reran them once, didn't work D:
